langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java

```diff
@@ -306,25 +316,34 @@ private Map<String, Object> prefillMetadata() {
     }
 
     private EmbeddingMatch<TextSegment> toEmbeddingMatch(Map<String, ?> item) {
-        Map<String, ?> additional = (Map<String, ?>) item.get(ADDITIONALS);
-        final Metadata metadata = new Metadata();
-        if (item.get(METADATA) != null && item.get(METADATA) instanceof Map) {
-            Map<String, ?> resultingMetadata = (Map<String, ?>) item.get(METADATA);
-            for (Map.Entry<String, ?> entry : resultingMetadata.entrySet()) {
-                if (entry.getValue() != null && !NULL_VALUE.equals(entry.getValue())) {
-                    metadata.add(entry.getKey(), entry.getValue());
-                }
-            }
+      Map<String, ?> additional = (Map<String, ?>) item.get(ADDITIONALS);
+      final Metadata metadata = new Metadata();
+      Map<String, ?> metadataMap = new HashMap<>();
+      if (metadataFieldName.isEmpty()) {
+        metadataMap = new HashMap<>(item);
+        // Remove text field from metadata if we store metadata in the root of the object
+        metadataMap.remove(textFieldName);
+      } else if (item.get(metadataFieldName) != null && item.get(metadataFieldName) instanceof Map) {
+        metadataMap = (Map<String, ?>) item.get(metadataFieldName);
+      }
+      if (metadataKeys != null && !metadataKeys.isEmpty()) {
+        metadataMap.keySet().retainAll(metadataKeys);
+      }
+      for (Map.Entry<String, ?> entry : metadataMap.entrySet()) {
+        if (entry.getValue() != null && !NULL_VALUE.equals(entry.getValue())) {
+          // TODO: Remove or replace use of deprecated method
+          metadata.add(entry.getKey(), entry.getValue());
```

Comment:
what about creating `Metadata` from `metadataMap` (with filtered out null values) instead of adding each metadata by key? then one does not need to use this deprecated method

Reply:
Done